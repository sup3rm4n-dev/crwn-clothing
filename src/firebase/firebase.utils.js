/* eslint-disable no-unused-vars */
/* eslint-disable no-use-before-define */
import firebase from 'firebase/app';
import 'firebase/auth';
import 'firebase/firestore';
// import 'firebase/performance';

// cSpell: disable
export const config = {
  apiKey: 'AIzaSyATNX2F21SsC33XuBZDOGyJwT1UgvCTPUg',
  authDomain: 'crwn-db-409fe.firebaseapp.com',
  databaseURL: 'https://crwn-db-409fe.firebaseio.com',
  projectId: 'crwn-db-409fe',
  storageBucket: 'crwn-db.appspot.com',
  messagingSenderId: '717530139246',
  appId: '1:717530139246:web:0fd962b63d08668f75347f',
};
// cSpell: enable

firebase.initializeApp(config);
// firebase.firestore().enablePersistence();

// const perf = firebase.performance();

export const createUserProfileDocument = async (userAuth, additionalData) => {
  if (!userAuth) return;

  const userRef = firestore.doc(`users/${userAuth.uid}`);

  const snapShot = await userRef.get();

  if (!snapShot.exists) {
    const { displayName, email } = userAuth;
    const createdAt = new Date();
    try {
      await userRef.set({
        displayName,
        email,
        createdAt,
        ...additionalData,
      });
    } catch (error) {
      console.log('error creating user', error.message);
    }
  }

  return userRef; // eslint-disable-line consistent-return
};

export const getUserCartRef = async userId => {
  const cartsRef = firestore.collection('carts').where('userId', '==', userId);
  const snapShot = await cartsRef.get();

  if (snapShot.empty) {
    const cartDocRef = firestore.collection('carts').doc();
    await cartDocRef.set({ userId, cartItems: [] });
    return cartDocRef;
  }
  return snapShot.docs[0].ref;
};

export const addCollectionAndDocuments = async (
  collectionKey,
  objectsToAdd
) => {
  const collectionRef = firestore.collection(collectionKey);

  const batch = firestore.batch();
  objectsToAdd.forEach(obj => {
    const newDocRef = collectionRef.doc();
    batch.set(newDocRef, obj);
  });

  return await batch.commit();
};

export const convertCollectionsSnapshotToMap = collections => {
  const transformedCollection = collections.docs.map(doc => {
    const { title, items } = doc.data();

    return {
      routeName: encodeURI(title.toLowerCase()),
      id: doc.id,
      title,
      items,
    };
  });

  return transformedCollection.reduce((accumulator, collection) => {
    accumulator[collection.title.toLowerCase()] = collection;
    return accumulator;
  }, {});
};

export const getCurrentUser = () => {
  return new Promise((resolve, reject) => {
    const unsubscribe = auth.onAuthStateChanged(userAuth => {
      unsubscribe();
      resolve(userAuth);
    }, reject);
  });
};

export const auth = firebase.auth();
export const firestore = firebase.firestore();

export const googleProvider = new firebase.auth.GoogleAuthProvider();
googleProvider.setCustomParameters({ prompt: 'select_account' });
export const signInWithGoogle = () => auth.signInWithPopup(googleProvider);

export default firebase;
