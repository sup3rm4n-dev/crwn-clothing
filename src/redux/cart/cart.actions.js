import CartActionTypes from './cart.types';
// import { createAction } from '../helpers/action.helpers';

const {
  TOGGLE_CART_HIDDEN,
  ADD_ITEM,
  REMOVE_ITEM,
  CLEAR_ITEM_FROM_CART,
  CLEAR_CART,
  SET_CART_FROM_FIREBASE,
} = CartActionTypes;

export const toggleCartHidden = () => ({
  type: TOGGLE_CART_HIDDEN,
});

// export const toggleCartHidden = createAction(TOGGLE_CART_HIDDEN);

export const addItem = item => ({
  type: ADD_ITEM,
  payload: item,
});

export const removeItem = item => ({
  type: REMOVE_ITEM,
  payload: item,
});

export const clearItemFromCart = item => ({
  type: CLEAR_ITEM_FROM_CART,
  payload: item,
});

export const clearCart = () => ({
  type: CLEAR_CART,
});

// export const setCartFromFirebase = cartItems =>
//   createAction(SET_CART_FROM_FIREBASE, cartItems);

export const setCartFromFirebase = cartItems => ({
  type: SET_CART_FROM_FIREBASE,
  payload: cartItems,
});
