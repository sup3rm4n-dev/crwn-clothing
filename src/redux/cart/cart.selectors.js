import { createSelector } from 'reselect';

import createDeepEqualSelector from '../deep-equal.selector';

const selectCart = state => state.cart;

const selectCartItems = createSelector([selectCart], cart => cart.cartItems);

const selectCartHidden = createSelector([selectCart], cart => cart.hidden);

const selectCartItemsCount = createDeepEqualSelector(
  [selectCartItems],
  cartItems =>
    cartItems.reduce(
      (accumalatedQuantity, cartItem) =>
        accumalatedQuantity + cartItem.quantity,
      0
    )
);

const selectCartTotal = createDeepEqualSelector([selectCartItems], cartItems =>
  cartItems.reduce(
    (acc, cartItem) => acc + cartItem.quantity * cartItem.price,
    0
  )
);

export {
  selectCartItems,
  selectCartHidden,
  selectCartItemsCount,
  selectCartTotal,
};
