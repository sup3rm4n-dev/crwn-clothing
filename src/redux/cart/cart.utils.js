/**
 * Simple helper function to add items to cart.
 * @param {array} cartItems Array of cartItems to search.
 * @param {object} cartItemToAdd Single new item to add to the cart.
 * @returns {array} New cart with added item.
 */
function addItemToCart(cartItems, cartItemToAdd) {
  const existingCartItem = cartItems.find(
    cartItem => cartItem.id === cartItemToAdd.id
  );

  if (existingCartItem) {
    return cartItems.map(cartItem =>
      cartItem.id === cartItemToAdd.id
        ? { ...cartItem, quantity: cartItem.quantity + 1 }
        : cartItem
    );
  }

  return [...cartItems, { ...cartItemToAdd, quantity: 1 }];
}

/**
 * Simple helper function to remove an item from cart.
 * @param {array} cartItems Array of cartItems to remove from.
 * @param {object} cartItemToAdd Single new item to remove from the cart.
 * @returns {array} Updated cart without removed item.
 */
function removeItemFromCart(cartItems, cartItemToRemove) {
  const existingCartItem = cartItems.find(
    cartItem => cartItem.id === cartItemToRemove.id
  );

  if (existingCartItem.quantity === 1) {
    return cartItems.filter(cartItem => cartItem.id !== cartItemToRemove.id);
  }

  return cartItems.map(cartItem => {
    return cartItem.id === cartItemToRemove.id
      ? { ...cartItem, quantity: cartItem.quantity - 1 }
      : cartItem;
  });
}

export { addItemToCart, removeItemFromCart };
