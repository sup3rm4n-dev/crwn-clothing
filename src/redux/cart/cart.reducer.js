/* eslint-disable no-param-reassign */
import produce from 'immer';

import { createEnum } from '../../helpers';
import CartActionTypes from './cart.types';
import { addItemToCart, removeItemFromCart } from './cart.utils';

const {
  TOGGLE_CART_HIDDEN,
  ADD_ITEM,
  REMOVE_ITEM,
  CLEAR_ITEM_FROM_CART,
  CLEAR_CART,
  SET_CART_FROM_FIREBASE,
} = CartActionTypes;

const initialState = createEnum({
  hidden: true,
  cartItems: [],
});

const cartReducer = produce((draft = initialState, action) => {
  switch (action.type) {
    case TOGGLE_CART_HIDDEN:
      draft.hidden = !draft.hidden;
      return draft;
    case ADD_ITEM:
      draft.cartItems = addItemToCart(draft.cartItems, action.payload);
      return draft;
    case REMOVE_ITEM:
      draft.cartItems = removeItemFromCart(draft.cartItems, action.payload);
      return draft;
    case CLEAR_ITEM_FROM_CART:
      draft.cartItems = draft.cartItems.filter(
        cartItem => cartItem.id !== action.payload.id
      );
      return draft;
    case CLEAR_CART:
      draft.cartItems = [];
      return draft;
    case SET_CART_FROM_FIREBASE:
      if (draft.cartItems?.length > 0) {
        draft.cartItems.push(action.payload);
      } else draft.cartItems = action.payload;
      return; // eslint-disable-line consistent-return
    default:
      return draft;
  }
});

export default cartReducer;
