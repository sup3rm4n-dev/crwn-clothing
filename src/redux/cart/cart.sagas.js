import { put, select } from 'redux-saga/effects';

import { getUserCartRef } from '../../firebase/firebase.utils';
import { selectCurrentUser } from '../user/user.selectors';
import { clearCart, setCartFromFirebase } from './cart.actions';
import { selectCartItems } from './cart.selectors';
import CartActionTypes from './cart.types';
import UserActionTypes from '../user/user.types';
import {
  createRootSaga,
  createTakeLatest,
  createTakeLeading,
} from '../helpers/saga.helpers';

const { SIGN_OUT_SUCCESS, SIGN_IN_SUCCESS } = UserActionTypes;
const { ADD_ITEM, REMOVE_ITEM, CLEAR_ITEM_FROM_CART } = CartActionTypes;

export function* clearCartOnSignOut() {
  yield put(clearCart());
}

export function* updateCartInFirebase() {
  const currentUser = yield select(selectCurrentUser);
  if (currentUser) {
    try {
      const cartRef = yield getUserCartRef(currentUser.id);
      const cartItems = yield select(selectCartItems);
      yield cartRef.update({ cartItems });
    } catch (error) {
      console.error(error);
    }
  }
}

export function* checkCartFromFirebase({ payload: user }) {
  try {
    const cartRef = yield getUserCartRef(user.id);
    const cartSnapshot = yield cartRef.get();
    yield put(setCartFromFirebase(cartSnapshot.data().cartItems));
  } catch (error) {
    // throw Error(`[saga] checkCartFromFirebase: ${error.message}`);
    console.warn(error.message);
  }
}

export const onSignOutSuccess = createTakeLatest(
  SIGN_OUT_SUCCESS,
  clearCartOnSignOut
);

export const onUserSignIn = createTakeLeading(
  SIGN_IN_SUCCESS,
  checkCartFromFirebase
);

const cartChangeActions = [ADD_ITEM, REMOVE_ITEM, CLEAR_ITEM_FROM_CART];
export const onCartChange = createTakeLatest(
  cartChangeActions,
  updateCartInFirebase
);

export const cartSagas = createRootSaga([
  onSignOutSuccess,
  onCartChange,
  onUserSignIn,
]);
