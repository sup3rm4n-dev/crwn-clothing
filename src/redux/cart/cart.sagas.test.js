import { takeLatest, put, all } from 'redux-saga/effects';
import tester from 'k-redux-saga-tester';
import UserActionTypes from '../user/user.types';
import { clearCart } from './cart.actions';
import {
  clearCartOnSignOut,
  onSignOutSuccess,
  checkCartFromFirebase,
  onUserSignIn,
  onCartChange,
  updateCartInFirebase,
  cartSagas,
} from './cart.sagas';
import { getUserCartRef } from '../../firebase/firebase.utils';
import CartActionTypes from './cart.types';

describe('on signout success saga', () => {
  it('should trigger on SIGN_OUT_SUCCESS', async () => {
    const generator = onSignOutSuccess();
    expect(generator.next().value).toEqual(
      takeLatest(UserActionTypes.SIGN_OUT_SUCCESS, clearCartOnSignOut)
    );
  });
});

describe('clear cart on signout saga', () => {
  it('should fire clearCart', () => {
    const generator = clearCartOnSignOut();
    expect(generator.next().value).toEqual(put(clearCart()));
  });
});

describe('check cart from firebase', () => {
  it('should fire getUserCartRef, calling with passed in user argument', () => {
    const mockUserID = { id: 'testing123123' };
    const mockUser = { payload: mockUserID };
    const generator = checkCartFromFirebase(mockUser);
    expect(generator.next().value).toEqual(getUserCartRef(mockUser.id));
  });
});

describe('on user sign in', () => {
  it('should fire UserActionType.SIGN_IN_SUCCESS, checkCartFromFirebase', () => {
    const generator = onUserSignIn();
    expect(generator.next().value).toEqual(
      takeLatest(UserActionTypes.SIGN_IN_SUCCESS, checkCartFromFirebase)
    );
  });
});

describe('on cart change', () => {
  it('should call cart actions', () => {
    const generator = onCartChange();
    const mockActions = [
      CartActionTypes.ADD_ITEM,
      CartActionTypes.REMOVE_ITEM,
      CartActionTypes.CLEAR_ITEM_FROM_CART,
    ];
    const mockNext = takeLatest(mockActions, updateCartInFirebase);
    expect(generator.next().value).toEqual(mockNext);
  });
});

describe('cartSaga snapshot', () => {
  it('should match snapshot', () => {
    const caller = tester(cartSagas);
    // const mockCalls = all([
    //   call(onSignOutSuccess),
    //   call(onCartChange),
    //   call(onUserSignIn),
    // ]);
    const mocks = {
      call: [
        () => [
          takeLatest(CartActionTypes.SIGN_OUT_SUCCESS, clearCartOnSignOut()),
          takeLatest(
            [
              CartActionTypes.ADD_ITEM,
              CartActionTypes.REMOVE_ITEM,
              CartActionTypes.CLEAR_ITEM_FROM_CART,
            ],
            updateCartInFirebase
          ),
          takeLatest(UserActionTypes.SIGN_OUT_SUCCESS, clearCartOnSignOut),
        ],
      ],
    };
    expect(caller(all(mocks))).toMatchSnapshot();
  });
});
