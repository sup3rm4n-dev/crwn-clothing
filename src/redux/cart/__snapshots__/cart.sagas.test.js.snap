// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cartSaga snapshot should match snapshot 1`] = `[Function]`;
